prow/git/publisher.go

```diff
@@ -0,0 +1,154 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package git
+
+import (
+	"fmt"
+	"net/url"
+)
+
+// RemoteResolver knows how to construct a remote URL for git calls
+type RemoteResolver interface {
+	Resolve() (string, error)
+}
+
+// StaticURLFor returns a generator that builds a static URL from the inputs.
+// This is useful when remote URLs never change, like when referring to an
+// authoritative clone URL for a repo.
+func StaticURLFor(host, org, repo string) func() (*url.URL, error) {
+	return func() (*url.URL, error) {
+		return &url.URL{Host: host, Path: fmt.Sprintf("%s/%s", org, repo)}, nil
+	}
+}
+
+// DynamicURLFor returns a generator that builds a dynamic URL from the inputs.
+// This is useful when remote URLs change, like when referring to a user's fork
+// of a repository.
+func DynamicURLFor(host, repo string, org func() (string, error)) func() (*url.URL, error) {
+	return func() (*url.URL, error) {
+		o, err := org()
+		if err != nil {
+			return nil, err
+		}
+		return &url.URL{Host: host, Path: fmt.Sprintf("%s/%s", o, repo)}, nil
+	}
+}
+
+// NewSimpleAuthResolver builds a dynamic resolver that knows how to
+// create simple auth URLs
+func NewSimpleAuthResolver(remote func() (*url.URL, error), username func() (string, error), token func() []byte) RemoteResolver {
+	return &simpleAuthResolver{
+		remote:   remote,
+		username: username,
+		token:    token,
+	}
+}
+
+// simpleAuthResolver builds URLs with simple auth credentials, resolved
+// dynamically. This structure is threadsafe.
+type simpleAuthResolver struct {
+	remote   func() (*url.URL, error)
+	username func() (string, error)
+	token    func() []byte
+}
+
+func (r *simpleAuthResolver) Resolve() (string, error) {
+	// copy the base endpoint to ensure we don't mutate it
+	remote, err := r.remote()
+	if err != nil {
+		return "", fmt.Errorf("could not resolve remote: %v", err)
+	}
+	name, err := r.username()
+	if err != nil {
+		return "", fmt.Errorf("could not resolve username: %v", err)
+	}
+	remote.User = url.UserPassword(name, string(r.token()))
+	return remote.String(), nil
+}
+
+// NewSSHResolver builds a resolver for SSH remotes
+func NewSSHResolver(host, repo string, org func() (string, error)) RemoteResolver {
```

Comment:
Nit: Can we maybe denote in the names that those are only good for github, i.E. `NewGitHubSSHResolver`?

Reply:
Why is this GitHub specific?

Reply:
Because other platforms don't necessarily have the org and repo concept